fence/man/fence_scsi.8
.TH fence_scsi 8

.SH NAME
fence_scsi - I/O fencing agent for SCSI persistent reservations

.SH SYNOPSIS
.B 
fence_scsi
[\fIOPTION\fR]...

.SH DESCRIPTION
fence_scsi is an I/O fencing agent which can be used with the SCSI
devices that support persistent reservations (SPC-2 or greater).

SCSI persistent reservations work by having each node in the cluster
register with the SCSI device. Registration is done using a unique key
(based on the node's IP address). Each node that will perform I/O
operations to the shared storage must register with the device. This
is done at system startup with the scsi_reserve init script. This
script will discover all clustered volumes as well as the underlying
SCSI device(s). Device discovery is done via the lvs command (see
lvs(8)) and is subject to any filtering rules defined in the lvm.conf
file.

After generating the node's unique key, the script will register the
node with the SCSI device(s) that were discovered. Once the node is
registered, the script will attempt to create a reservation. Unlike
registrations, of which there are multiple registrants (one for each
node in the cluster), there is only one reservation holder. If a
reservation does not already exist for a device, the script will
create a reservation using the node's unique key.

It is important to distinguish between registrations and
reservations. As mentioned above, each node that will perform I/O
operations to the SCSI device must register. As a result, there will
be multiple registrations for a given SCSI device. In contrast, there
can only be one reservation per SCSI device. It is not important which
node holds the reservation. The reservation simply tells the device
how the registrants are allowed to access the device. For our
purposes, the reservation type is "write exclusive, registrants only".
With this reservation type, only registered nodes will be able to
write to the device.

When the cluster must fence a node, it simply revokes a node's
registration, or "unregisters" the node. This operation also uses the
node's unique key. By deriving the unique key based on the errant
node's IP address, the cluster can "unregister" the key. As a
result, the errant node will no longer be able to write to the SCSI
device.

Note that the node that holds the reservation for a device must also
be registered with that device. When the situation arises where the
node that is being fenced is also the reservation holder, the
reservation must be moved. This is handled by using the
"preempt-and-abort command" which will transfer the reservation from
the node that is being fenced to the node that is performing the
fencing. This operation will maintain the reservation while
"unregistering" the node being fenced.

At system shutdown, the scsi_reserve script will attempt to
"unregister" the node from all devices. The exception is when the
node happens to be the reservation holder. In this case, the script
does nothing, due to the fact that there may be other nodes using the
device and the reservation must remain intact.

fence_scsi accepts options on the command line as well as from stdin.
fenced sends parameters through stdin when it execs the agent.  fence_scsi
can be run by itself with command line options.  This is useful for testing 
and for turning outlets on or off from scripts.

.SH OPTIONS
.TP
\fB-n\fP \fInode\fR
Name of the node to be fenced.
.TP
\fB-h\fP
Print out a help message describing available options, then exit.
.TP
\fB-v\fP
Verbose output.
.TP
\fB-V\fP
Print out a version message, then exit.

.SH STDIN PARAMETERS
.TP
\fIagent = < param >\fR
This option is used by fence_node(8) and is ignored by fence_scsi.
.TP
\fInodename = < hostname | ip >\fR
Name of the node to be fenced.
.TP
\fIself = < nodename >\fR
Name of the node that will perform the fencing operation.
.TP
\fIverbose = < param >\fR
Verbose output.

.SH LIMITATIONS
The fence_scsi fencing agent requires a minimum of three nodes in the
cluster to operate.  For SAN devices connected via fiber channel,
these must be physical nodes.  SAN devices connected via iSCSI may use
virtual or physical nodes.  In addition, fence_scsi cannot be used in
conjunction with qdisk.

.SH SEE ALSO
fence(8), fence_node(8), sg_persist(8), lvs(8), lvm.conf(5)


fence/agents/scsi/fence_scsi.pl
#!/usr/bin/perl

use Getopt::Std;
use IPC::Open3;
use POSIX;

my $ME = $0;

END {
  defined fileno STDOUT or return;
  close STDOUT and return;
  warn "$ME: failed to close standard output: $!\n";
  $? ||= 1;
}

$_ = $0;
s/.*\///;
my $pname = $_;
my @device_list;

#BEGIN_VERSION_GENERATION
$RELEASE_VERSION="";
$REDHAT_COPYRIGHT="";
$BUILD_DATE="";
#END_VERSION_GENERATION

sub usage
{
    print "Usage\n";
    print "\n";
    print "$pname [options]\n";
    print "\n";
    print "Options:\n";
    print "  -n <node>		ip address or hostname of node to fence\n";
    print "  -h			usage\n";
    print "  -u			unfence\n";
    print "  -v			verbose\n";
    print "  -V			version\n";

    exit 0;
}

sub version
{
    print "$pname $RELEASE_VERSION $BUILD_DATE\n";
    print "$REDHAT_COPYRIGHT\n" if ($REDHAT_COPYRIGHT);
}

sub fail_usage
{
    ($msg) = @_;

    print STDERR $msg."\n" if $msg;
    print STDERR "Please use '-h' for usage.\n";

    exit 1;
}

sub check_sg_persist
{
    my ($in, $out, $err);
    my $cmd = "sg_persist -V";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "unable to execute sg_persist.\n" if ($?>>8);

    close ($in);
    close ($out);
    close ($err);
}

sub get_cluster_id
{
    my ($in, $out, $err);
    my $cmd = "cman_tool status";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "unable to execute cman_tool.\n" if ($?>>8);

    my $cluster_id;

    while (<$out>) {
	chomp;

	my ($name, $value) = split(/\s*:\s*/, $_);

	if ($name eq "Cluster Id") {
	    $cluster_id = $value;
	    last;
	}
    }

    close ($in);
    close ($out);
    close ($err);

    return $cluster_id;
}

sub get_node_id
{
    ($node) = @_;

    my ($in, $out, $err);
    my $cmd = "ccs_tool query /cluster/clusternodes/clusternode[\@name=\\\"$node\\\"]/\@nodeid";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "Unable to execute ccs_tool.\n" if ($?>>8);

    while (<$out>) {
        chomp;
        $node_id = $_;
    }

    close ($in);
    close ($out);
    close ($err);

    return $node_id;
}

sub get_host_name
{
    my ($in, $out, $err);
    my $cmd = "cman_tool status";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "unable to execute cman_tool.\n" if ($?>>8);

    my $host_name;

    while (<$out>) {
	chomp;

	my ($name, $value) = split(/\s*:\s*/, $_);

	if ($name eq "Node name") {
	    $host_name = $value;
	    last;
	}
    }

    close ($in);
    close ($out);
    close ($err);

    return $host_name;
}

sub get_node_name
{
    return $opt_n;
}

sub get_key
{
    ($node) = @_;

    my $cluster_id = get_cluster_id;
    my $node_id = get_node_id($node);

    if ($node_id == 0) {
	die "unable to determine nodeid for node '$node'\n";
    }

    my $key = sprintf "%.4x%.4x", $cluster_id, $node_id;

    return $key;
}

sub get_options_stdin
{
    my $opt;
    my $line = 0;

    while (defined($in = <>)) {

	$_ = $in;
	chomp;

	## strip leading and trailing whitespace
	##
	s/^\s*//;
	s/\s*$//;

	## skip comments
	##
	next if /^#/;

	$line += 1;
	$opt = $_;

	next unless $opt;

	($name, $value) = split(/\s*=\s*/, $opt);

	if ($name eq "")
	{
	    print STDERR "parse error: illegal name in option $line\n";
	    exit 2;
	}
	elsif ($name eq "agent")
	{
	    ## ignore this
	}
	elsif ($name eq "node")
	{
	    $opt_n = $value;
	}
	elsif ($name eq "nodename")
	{
	    $opt_n = $value;
	}
	elsif ($name eq "action")
	{
	    ## if "action=on", the we are performing an unfence operation.
	    ## any other value for "action" is ignored. the default is a
	    ## fence operation.
	    ##
	    if ($value eq "on") {
		$opt_u = $value;
	    }
	}
    }
}

sub get_scsi_devices
{
    my ($in, $out, $err);

    my $cmd = "vgs --config 'global { locking_type = 0 }'" .
              "    --noheadings --separator : -o vg_attr,pv_name 2> /dev/null";

    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "unable to execute vgs.\n" if ($?>>8);

    while (<$out>) {
	chomp;

	my ($attrs, $dev) = split(/:/, $_);

	if ($attrs =~ /.*c$/) {
	    $dev =~ s/\(.*\)//;
	    push(@device_list, $dev);
	}
    }

    close ($in);
    close ($out);
    close ($err);
}

sub create_registration
{
    my ($key, $dev) = @_;

    my ($in, $out, $err);
    my $cmd = "sg_persist -n -d $dev -o -I -S $key";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "unable to create registration.\n" if ($?>>8);

    close ($in);
    close ($out);
    close ($err);
}

sub create_reservation
{
    my ($key, $dev) = @_;

    my ($in, $out, $err);
    my $cmd = "sg_persist -n -d $dev -o -R -K $key -T 5";
    my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

    waitpid($pid, 0);

    die "unable to create reservation.\n" if ($?>>8);

    close ($in);
    close ($out);
    close ($err);
}

sub fence_node
{
    my $host_name = get_host_name;
    my $host_key = get_key($host_name);

    my $node_name = get_node_name;
    my $node_key = get_key($node_name);

    my ($in, $out, $err);

    foreach $dev (@device_list)
    {
	## check that the key we are attempting to remove
	## is actually registered with the device. if the key is not
	## registered, there is nothing to do for this device.
	##
	system ("sg_persist -n -d $dev -i -k | grep -qiE \"^[[:space:]]*0x$key\"");

	if (($?>>8) != 0) {
	    next;
	}

	if ($host_key eq $node_key) {
	    ## this sg_persist command is for the case where you attempt to
	    ## fence yourself (ie. the local node is the same at the node to
	    ## be fence). this will not work if the node is the reservation
	    ## holder, since you can't unregister while holding the reservation.
	    ##
	    my $cmd = "sg_persist -n -d $dev -o -G -K $host_key -S 0";
	}
	else {
	    ## this sg_persist command will remove the registration for $host_key.
	    ## the local node will also become the reservation holder, regardless
	    ## of which node was holding the reservation prior to the fence operation.
	    ##
	    my $cmd = "sg_persist -n -d $dev -o -A -K $host_key -S $node_key -T 5";
	}

	my $pid = open3($in, $out, $err, $cmd) or die "$!\n";

	waitpid($pid, 0);

	die "unable to execute sg_persist.\n" if ($?>>8);

	close ($in);
	close ($out);
	close ($err);
    }
}

sub unfence_node
{
    my $host_name = get_host_name;
    my $host_key = get_key($host_name);

    foreach $dev (@device_list)
    {
	create_registration ($host_key, $dev);

	## check to see if a reservation already exists.
	## if no reservation exists, this node/key will become
	## the reservation holder.
	##
	system ("sg_persist -n -d $dev -i -r | grep -qiE \"^[[:space:]]*Key=0x\"");

	if (($?>>8) != 0) {
	    create_reservation ($host_key, $dev);
	}
    }
}

if (@ARGV > 0) {

    getopts("hn:uvV") || fail_usage;

    usage if defined $opt_h;
    version if defined $opt_V;

    if (!defined $opt_u) {
	fail_usage "No '-n' flag specified." unless defined $opt_n;
    }

} else {

    get_options_stdin();

}

## get a list of scsi devices. this call will build a list of devices
## (device_list) by querying clvm for a list of devices that exist in
## volume groups that have the cluster bit set.
##
get_scsi_devices;

if ($opt_u) {
    unfence_node;
} else {
    fence_node;
}
